sycl/source/detail/platform_impl.cpp

```diff
@@ -303,9 +304,20 @@ static void filterDeviceFilter(vector_class<RT::PiDevice> &PiDevices,
   if (!FilterList)
     return;
 
+  // remember the last backend that has gone through this filter function
+  // to assign a unique device id number across platforms that belong to
+  // the same backend. For example, opencl:cpu:0, opencl:acc:1, opencl:gpu:2
+  static backend lastBackend = backend::all;
   backend Backend = Plugin.getBackend();
   int InsertIDx = 0;
-  int DeviceNum = 0;
+  // DeviceNums should be given consecutive numbers across platforms.
+  // So, we keep the device num for the successive calls to this function.
+  static int DeviceNum = 0;
+  if (lastBackend != Backend) {
```

Comment:
This seems to assume that all platforms/devices from the same backend are going together in the same group, which is too fragile. I think you need to remember last device number in each backend.

Reply:
Yes, platforms are added together from a Plugin. 
For example, when OpenCL backend is added, we add all supported platforms from OpenCL together.
So, they will be always grouped together.
I also thought about remembering the last device number for each backend, but it is not correct if the user calls get_devices() multiple times.
We should always reset to zero for different backend.

Reply:
> So, they will be always grouped together.

That's implementation detail, and what I call fragile. Please think how to make it more robust.

Reply:
I agree it is not a good design, but I can't come up with any better solution.
As I mentioned before, all platforms are come together, and it is impossible to scatter platforms across different backends.
I will take your suggestion if you have a good idea to solve this issue without utilizing the implementation details.

Reply:
I think we should make the .get_devices to always run only once. It should obtain, enumerate, filter, cache what's needed such that subsequent calls just return that.

Reply:
Hmm...  Sorry I don't follow. Please help me understand your suggestion.
I am not sure how the device cache can solve the issue of assigning unique device id per backend across different platforms.
The issue we are talking about here is the VERY FIRST TIME we assign the device id unique across different platforms without utilizing the implementation details.

Reply:
I am also talking about making code thread-safe and removing the redundancy of reaching out to plugins every time `platform_impl::get_devices()` is called. The way I envision it is that there should be code that performs **once**:

``` C++
for (every backend)
  for (every platform)
    for (every device) {
      // 1. filter
      // 2. enumerate
      // 3. cache
    }
```
Current code needs to be restructured so that the enumeration is done in the same place where platforms/devices are iterated. Subsequent calls to `.get_platforms()` and `.get_devices()` should grab everything from the cache.

Reply:
Yes, I agree that if we use device cache, we don't need to recalculate the device numbers.
However, that's just an optimization and it won't help solve the issue you raised.
Actually, if we can get it right very first time, it does not matter whether we have device cache or not.
The problem is to give unique device ids across multiple platforms at the very first time when creating the cache ONCE.
The problem still did not go away If multiple threads are calling get_devices() during the cache creation time (ONCE).
I am still not clear how we can solve this issue without utilizing the implementation detail.

Reply:
1. If you re-structure code as I suggested then you can do stable enumeration.
2. The MT problem would also be solved (other threads won't run that code simultaneously, they will wait for populated cache)

Reply:
We already have a global cache for platforms and devices in SYCL RT, and never creates duplicate platforms and devices.
In other words, we do exactly what you suggested.
The very first call to get_devices() will create them in cache ONCE, reuse it over and over again.
But it does not keep the device id in cached devices.
If I add this device id info in the device class, does it solve your concern?